### DISCUSSION TITLE
How to check specific users (friends) online/offline presence in React Native Firebase

### DISCUSSION THREAD

**Discussion by owaisulwara** [invertase/react-native-firebase#4191] [Q&A] [answered]
I'm working on an app where I need to show the **friend's online/offline presence** _( not all users )_, as you can see in the picture below. If I listen to users presence method then what will I do if the users are more than 1000? it gets the data downloaded on every change which I don't want because all of the users are not in friends list.

![image](https://user-images.githubusercontent.com/28351659/91864674-ee3ef700-ec89-11ea-8c31-5fba3e9e4495.png)

What if I add multiple listeners for every friends then how could it be efficient?

Thanks

**Comment by andersonaddo** [comment]
Well the first question to ask is
Are you using Realtime Database or Firestore?

**Reply by owaisulwara** [reply]
I'm using firestore.

**Reply by andersonaddo** [reply]
Well then there are a number of ways to achieve this.
You could have every user in the `user` collection have an array of uids corresponding to their friends.
Then to get something similar to what you have above you can just query the `users` collection for docs that have the user's uid in that `friends` array. One could argue that this is the best way since presence information for any user only has to be updated in their doc that's in the users collection

Or another way is for every user to have their own `friends` collection, which essentially contains copies of docs corresponding to their friends. Then, to get what you have above, you just query straight from that user's `friends` collection. The only problem with this is that user presence data may have to be written (per user) in all their duplicate docs that are in the multiple `friends` collections.

**Reply by owaisulwara** [reply]
@andersonaddo please read the question. I added online/offline. Actually I'm concerned about **real-time user's friends online/offline presence**. Currently I acheived it by adding multiple listeners. Each listener for each friend. Please let me know if there is any better way to acheive this.

**Reply by andersonaddo** [reply]
If you're able to get all online/offline information via one listener, then that's better than adding several listeners. Even though, on every change, the listener will give you back the whole a new version of the while dataset, the SDK code actually only gets the deltas from the server. It then merges the deltas with a cached version of the whole dataset and serves you that on the client side.

Also, for the record, you never explicitly mention that you already implemented this. You phrased your question like you _wanted_ to firgure out a way to do this, and that image below is an example of what you want.
But whatever, maybe just me.

**Reply by andersonaddo** [reply]
I would suggest watching this series 
https://www.youtube.com/watch?v=v_hR4K4auoQ&feature=emb_logo

**Comment by andersonaddo** [comment]
Also you should learn about pagination

**Reply by owaisulwara** [reply]
I read about pagination but how can it help me to acheive the goal?

**Comment by mikehardy** [comment, marked as answer]
I'm not sure this is react-native-firebase specific at all. The same problem would apply for presence detection for any firebase product. They have support channels upstream and a heavy presence in learning activities, @owaisulwara have you thought about contacting Google Firebase team through their official education channels and seeing if they have any applicable case studies or would do one? They do a few videos a week and this is an interesting problem - if they have not already tackled it they might be interested.

**Comment by cdg720** [comment]
@owaisulwara did you find a good solution? I have the same problem.

**Comment by federico-aguilar** [comment]
Does anyone have an answer to efficiently setup a presence system? I've only encountered very few posts regarding a firestore presence system and they don't have answers but generic approaches: 
https://stackoverflow.com/questions/67656073/how-do-people-build-a-friend-presence-system-not-a-global-user-presence-system
https://stackoverflow.com/questions/65798427/efficient-and-scalable-social-presence-with-firebase-realtime-database
